src/main/java/com/google/api/codegen/metacode/InitCodeNode.java

```diff
@@ -431,6 +427,38 @@ private static Name getChildSuggestedName(
     }
   }
 
+  /**
+   * For a read-from-file node, we set up a simple child node to assign the file name to a local
+   * variable (e.g., String fileName = "file_name.jpg"). If sample_argument_name is specified, the
+   * local variable would honor the configuration. If not, the name of the local variable would
+   * default to "file_name", and would be picked automatically by symbolTable if "file_name" is
+   * already in use. This allows us to pass in the file name as a sample function argument.
+   */
+  private void setupReadFileNode(InitCodeContext context) {
+    Preconditions.checkArgument(
+        children.isEmpty(), "Can only configure leaf node to read from file.");
+    setLineType(InitCodeLineType.ReadFileInitLine);
+    Name childIdentifier;
+    if (sampleParamConfig.isSampleArgument()) {
+      Name name = Name.anyLower(sampleParamConfig.sampleArgumentName());
+      Preconditions.checkArgument(
+          !context.symbolTable().contains(name),
+          "sample_argument_name \"%s\" is already in use.",
+          sampleParamConfig.sampleArgumentName());
+      childIdentifier = context.symbolTable().getNewSymbol(name);
+    } else {
+      childIdentifier = context.symbolTable().getNewSymbol(Name.anyLower(FILE_NAME_KEY));
+    }
+    InitCodeNode child =
+        new InitCodeNode(FILE_NAME_KEY, InitCodeLineType.SimpleInitLine, initValueConfig);
```

Comment:
Re: "Just double-checking: Having FILE_NAME_KEY does not preclude us from having multiple filenames in a single sample, correct?"
- No, it shouldn't. it's within the scope of a ReadFileNode, and for now there should be only one child node for a ReadFIleNode. However when we implement comma-separated file names, we will need to modify this (probably by having a `FileNameKeyFactory` generate keys?)

Re: "Reading over the comments in getKey(), how will you handle things when we have an iterable of filenames? In that case, the key should be the map or list index, but that's different than FILE_NAME_KEY, right?"
- Ah I see the confusion here. `FILE_NAME_KEY` is a constant string `"file_name"`, but is used for both the key to identify the child node and the name of the local variable to store the actual file name. That being said, it's probably better to have two constants `FILE_NAME_KEY` and `FILE_NAME_LOCAL_VAR_NAME`, even if the both of them will have the value of `"file_name"`? For an iterable of filenames, we may need a factory as mentioned above.

Re:"what happens if we have a map that happens to have the key file_name..."
- We won't screw up setting up the tree because a ReadFileNode shouldn't have any children before we adding the child node whose key is `file_name`.
- We won't screw up transforming the tree to views because we only treat the key `file_name` differently when the `lineType` of the parent node is `ReadFileLineType`. If the parent is a map node, we render it as a map regardless of the keys of it's children.
- IIUC, the key is only used to identify children nodes from the context of the parent node, the children nodes have their own type specified (the field `InitCodeNode.lineType`).
- However, I do think it makes a lot of sense to make the statics start with a illegal character. I'll change `FILE_NAME_KEY` in this PR but we probably should handle `root` in a separate PR.

Reply:
Ah, thanks. I think I get it now. Let me check:

1. When you refer to `ReadFileNode`, you mean an `InitCodeNode` with `type==ReadFileInitLine`, correct?
2. When a parameter is specified with an comma-separated list of filenames, do we need to have a key factory or can we simply use a numeric-ish index?
3. `FILE_NAME_KEY` vs `FILE_NAME_LOCAL_VAR_NAME`: Yes, let's disambiguate the uses, even if the values we employ are the same. BTW, for the latter, maybe `LOCAL_DEFAULT_FILENAME`
4. iterable of filenames: I was referring to the case where a user wants to individually specify values of a map, and those values are to be treated by us as filenames. Eg the user sample config specifies
```
images["mickey"]="mickey.jpg"
images["donald"]="donald.jpg"
```
In that case, what would the init code node structure be? IIUC (and I may not be!):
  a. each child of `images` is stored under a store-key and internally has an internal-key, and both of those are the same
  b. according to the comment in `getKey()` the internal-keys reference the location in the parent, so the internal-keys for both elements above would be "mickey" and "donald"
  c. Each of the `InitCodeNode`s for `images["mickey"]` and `images["donald"]` would have `type==ReadFileInitLine`, and _each_ of those would in turn have one child (with store-key==internal-key==`FILE_NAME_KEY`) encoding the filename. If that's how it works, that addresses my previous question
5. Having the statics start with an illegal character makes sense.
6. Even without the illegal start character: in the case of a user-specified map where one of the keys happens to be equal to the value of `FILE_NAME_KEY` (and the corresponding value is intended to be just a string), we would be able to disambiguate that from a file-input-parameter because of `type!=ReadFileInitLine`. Did I understand correctly?

Reply:
Yes! All of them are what the code is intended to.